src/features/call/components/SurveyAccordion.tsx

```diff
@@ -0,0 +1,154 @@
+  AccordionDetails,
+  Typography,
+  Box,
+  Button,
+} from '@mui/material';
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
+
+import ZUILogoLoadingIndicator from 'zui/ZUILogoLoadingIndicator';
+import useSurveyElements from 'features/surveys/hooks/useSurveyElements';
+import useCreateSurveySubmission from 'features/surveys/hooks/useCreateSurveySubmission';
+import SurveyText from './Survey/SurveyText';
+import SurveyOptions from './Survey/SurveyOptions';
+import { ZetkinCallTarget } from '../types';
+import {
+  ELEMENT_TYPE,
+  ZetkinSurveyExtended,
+  ZetkinSurveyOptionsQuestionElement,
+  ZetkinSurveyQuestionResponse,
+  ZetkinSurveyTextQuestionElement,
+} from 'utils/types/zetkin';
+
+interface SurveyAccordionProps {
+  orgId: number;
+  survey: ZetkinSurveyExtended;
+  target: ZetkinCallTarget;
+}
+
+const SurveyAccordion: FC<SurveyAccordionProps> = ({
+  orgId,
+  survey,
+  target,
+}) => {
+  const elements = useSurveyElements(orgId, survey.id).data;
+  const { addSubmission } = useCreateSurveySubmission(orgId, survey.id);
+  const [responses, setResponses] = useState<{
+    [key: number]: { options?: number[]; response?: string };
+  }>({});
+
+  const handleResponseChange = (
+    questionId: number,
+    value: number[] | string | number
+  ) => {
+    setResponses((prevResponses) => ({
+      ...prevResponses,
+      [questionId]: Array.isArray(value)
+        ? { options: value }
+        : { response: String(value) },
+    }));
+  };
+
+  const handleFormSubmission = () => {
+    const signature = target.id;
+    const formResponses: {
+      [key: number]: { options?: number[]; response?: string };
+    } = responses;
+
+    const mappedResponses: ZetkinSurveyQuestionResponse[] = Object.keys(
+      formResponses
+    )
+      .map((key) => {
+        const questionId = parseInt(key);
+        const response = formResponses[questionId];
+
+        if (response.response) {
+          return {
+            question_id: questionId,
+            response: response.response,
+          };
+        } else if (response.options) {
+          return {
+            options: response.options,
+            question_id: questionId,
+          };
+        }
+
+        return null;
+      })
+      .filter(
+        (response): response is ZetkinSurveyQuestionResponse =>
+          response !== null
+      );
+
+    addSubmission({
+      responses: mappedResponses,
+      signature,
+    });
+  };
+
+  return (
+    <Suspense fallback={<ZUILogoLoadingIndicator />}>
+      <Accordion>
+        <AccordionSummary expandIcon={<ExpandMoreIcon />}>
+          <Typography>{survey.title}</Typography>
+        </AccordionSummary>
+        <AccordionDetails>
+          <Box mb={2}>
+            <Typography variant="h6">{survey.title}</Typography>
+            <Typography>{survey.info_text}</Typography>
+          </Box>
+          {elements &&
+            elements.length > 0 &&
+            elements.map((element) => {
```

Comment:
I would consider using `<SurveyForm/>`, or at least `<SurveyElements/>` here instead of rendering the element list manually. Is that possible?

Reply:
I think I can't use either of those components without modifying them, because with those I can't handle the responses or store the data in the state for later submit it from the `SurveyAccordion`. Let me know if I'm mistaken and there's a way to do it.

Reply:
I understand where you're coming from. You've built this UI using state and `onChange()` handlers, but that should not be necessary. I would look into using a `<form>` around the `SurveyElements` and listen to the `onSubmit()` of that form to get all of the data from all of the fields. The logic for extracting it from a `FormData` has already been created once, so submitting it here should not require too much bespoke stuff.

Reply:
I've fixed it in commit: [450dbf9](https://github.com/zetkin/app.zetkin.org/pull/2607/commits/450dbf9645133a0833d84271da03978224e1bf34)